File: solution/binary_search/2470/main.cpp
// Authored by : fltcy2039
// Co-authored by : -
// Link : http://boj.kr/fd177f1f686d409484f231c8b81ca146

#include <bits/stdc++.h>
using namespace std;
int arr[100005];

int main(){
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    
    //input & init
    int N; cin >> N;
    for(int i = 0 ; i < N; i++) cin >> arr[i];
    
    int min;
    int start, end, answer_start, answer_end;
    
    //process
    sort(arr, arr + N);
    
    //답 후보 인덱스 answer_start, answer_end
    //갱신을 위해 이동하는 인덱스 start, end
    answer_start = start = 0;
    answer_end = end = N - 1;
    min = abs(arr[start]+arr[end]);
        
    int sum = min;
    while(start < end){
        sum = arr[start] + arr[end];
        if(min > abs(sum)){
            min = abs(sum);
            answer_start = start;
            answer_end = end;
        }
        
        //절댓값을 줄이는 방향으로 인덱스 이동
        if(sum < 0) start++;
        else end--;
    }
    
    //output
    cout << arr[answer_start] << " "<< arr[answer_end] << "\n";
    return 0;
}

File: solution/binary_search/2470/main.py
# Authored by : cieske
# Co-authored by : -
# Link : http://boj.kr/c487c1f5e7f34df0aaeba6a14fb6b2a5
from bisect import bisect_left
import sys
def input():
    return sys.stdin.readline().rstrip()

n = int(input())
lst_pos, lst_neg = [], []
input_lst = list(map(int, input().split()))
for x in input_lst:
    if x > 0: lst_pos.append(x)
    else: lst_neg.append(x)

lst_pos.sort()
lst_neg.sort()

if not lst_pos: print(*lst_neg[-2:]) # 양수가 없는 경우
elif not lst_neg: print(*lst_pos[:2]) # 음수가 없는 경우
else:
    tmp = 2000000001 #최악 케이스+1
    if len(lst_neg) > 1: # 음수만 가지고 답일 수 있음
        if tmp > abs(sum(lst_neg[-2:])):
            tmp = abs(sum(lst_neg[-2:]))
            sol = lst_neg[-2:]

    if len(lst_pos) > 1: # 양수만 가지고 답일 수 있음
        if tmp > abs(sum(lst_pos[:2])):
            tmp = abs(sum(lst_pos[:2]))
            sol = lst_pos[:2]  

    for num in lst_neg:
        idx = bisect_left(lst_pos, -num)
        if idx == len(lst_pos): idx -= 1 # 마지막 인덱스
        elif idx == 0: pass # 처음 인덱스
        elif abs(num+lst_pos[idx]) > abs(num+lst_pos[idx-1]): # 둘 중 최적 선택
            idx -= 1

        if tmp > abs(num+lst_pos[idx]): # 필요 시 업데이트
            tmp = abs(num+lst_pos[idx])
            sol = sorted([num, lst_pos[idx]])

    print(*sol)
